metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/execution/BlockExecutor.scala

```diff
@@ -0,0 +1,304 @@
+              // If a block fails, return what we managed to do so far,
+              // so we can re-attempt it next time if the block is still
+              // available in the storage.
+              tracers
+                .error(s"Error executing block $blockHash", ex)
+
+            case Right(None) =>
+              // Either the block couldn't be found, or the last executed
+              // hash changed, but something suggests that we should not
+              // try to execute more of this batch.
+              nextBlockHashes.traverse(tracers.executionSkipped(_)).void
+
+            case Right(Some(nextLastExecutedBlockHash)) =>
+              loop(nextBlockHashes, nextLastExecutedBlockHash)
+          }
+      }
+
+    loop(newBlockHashes, lastExecutedBlockHash)
+  }
+
+  /** Execute the next block in line and update the view state.
+    *
+    * The last executed block hash is only updated if the application
+    * indicates that it has persisted the results, and if no other
+    * changes have been made to it outside this loop. The execution
+    * result carries the new last executed block hash to use in the
+    * next iteration, or `None` if we should abandon the execution.
+    */
+  private def executeBlock(
+      blockHash: A#Hash,
+      commitQC: QuorumCertificate[A],
+      commitPath: NonEmptyList[A#Hash],
+      lastExecutedBlockHash: A#Hash
+  ): F[Option[A#Hash]] = {
+    assert(commitPath.head == blockHash)
+    assert(commitPath.last == commitQC.blockHash)
+
+    storeRunner.runReadOnly {
+      blockStorage.get(blockHash)
+    } flatMap {
+      case None =>
+        tracers.executionSkipped(blockHash).as(none)
+
+      case Some(block) =>
+        for {
+          isPersisted <- appService.executeBlock(block, commitQC, commitPath)
+          _           <- tracers.blockExecuted(blockHash)
+
+          maybeLastExecutedBlockHash <-
+            if (!isPersisted) {
+              // Keep the last for the next compare and set below.
+              lastExecutedBlockHash.some.pure[F]
+            } else {
+              // Check that nothing else changed the view state,
+              // which should  be true as long as we use the semaphore.
+              // Otherwise it would be up to the `ApplicationService` to
+              // take care of isolation, and check that the block being
+              // executed is the one we expected.
+              setLastExecutedBlockHash(blockHash, lastExecutedBlockHash).map {
+                case true  => blockHash.some
+                case false => none
+              }
+            }
+        } yield maybeLastExecutedBlockHash
+    }
+  }
+
+  /** Replace the state we have persisted with what we synced with the federation.
+    *
+    * Prunes old blocks, the Commit Q.C. will be the new root.
+    */
+  private def fastForwardStorage(
+      block: A#Block,
+      canPrune: Boolean
+  ): F[Unit] = {
+    val blockHash = Block[A].blockHash(block)
+
+    val prune = for {
```

Comment:
1. Why are we not using `blockStorage.pruneNonDescendants` here? Is it because there may be no link between the root and the `blockhash`? 
2. What about the safe deletion order?
3. Should we maybe put this into `BlockStorage` as an alternative pruning method so that the difference in use cases can be more easily spotted?

Reply:
That got me scratching my head, I should add a comment. 

1. You're exactly right, when `syncState` is invoked we can expect that there will be a gap between what we have in the block storage and the block to which we just synced, so there is (probably) no way to traverse to the root from the new block. That's why I added the root block hash explicitly to the view state, so we can go from the other direction. The `skip` is used purely as a safeguard.
2. Arguably at this point we don't care about the deletion order because we assume we are dealing with a forest, and we just want to purge the old tree and keep the new (single) tree. It will be executed as one batch against RocksDB. 
3. The thing is that the block store doesn't store the root, we rely on the view store here to figure out where to go from. We discussed earlier that the block store _can_ store a forest, and it's up to the user to keep track of pointers into each tree. The view store knows a bit more, we know we only care about this one tree. We could move a method that just purges all descendants of a certain block, but there I would try to be more careful about the deletion order as I could not assume the block is the root... maybe a method that goes up to the root and kills the root, a `purgeTree`. Should we do that?

Reply:
Added in https://github.com/input-output-hk/metronome/pull/39/commits/0b8b29d90c4c45ee2556482664880d0bc0d210cf
Thanks for the suggestion!

Reply:
I like it better that way :+1: